File: 2022/day-4.rb
input = File.read('day-4-input.txt').split("\n")


result1 = input.select do |l|
  r1, r2 = l.split(",")
  start1, end1 = r1.split("-").map(&:to_i)
  start2, end2 = r2.split("-").map(&:to_i)

  (start1..end1).cover?(start2..end2) || (start2..end2).cover?(start1..end1)
end

puts "Part 1: #{ result1.length }"

result2 = input.select do |l|
  r1, r2 = l.split(",")
  start1, end1 = r1.split("-").map(&:to_i)
  start2, end2 = r2.split("-").map(&:to_i)

  (start1..end1).include?(start2) || (start1..end1).include?(end2) || (start2..end2).include?(start1) || (start2..end2).include?(end1)
end
# puts result2
puts "Part 2: #{ result2.length }"

File: 2022/day-4.ex
input = File.read!("day-4-input.txt")

defmodule AoC do
  def parseLine(line) do
    line
    |> String.split(",") # "1-2,3-4" -> ["1-2", "3-4"]
    |> Enum.map(fn s ->  # ["1-2", "3-4"] -> [[1, 2], [3, 4]]
      s
      |> String.split("-")
      |> Enum.map(&String.to_integer/1)
    end)
  end

  def cover([[start1, end1], [start2, end2]]) do
    start1 <= start2 && start2 <= end1 && start1 <= end2 && end2 <= end1
  end

  def overlap([[start1, end1], [start2, end2]]) do
    !Range.disjoint?(start1..end1, start2..end2)
  end
end

data =
  input
  |> String.trim()
  |> String.split("\n")
  |> Enum.map(&AoC.parseLine/1)

result1 =
  data
  |> Enum.filter(fn pairs -> AoC.cover(pairs) || AoC.cover(Enum.reverse(pairs)) end)
  |> Enum.count()

IO.inspect(result1)
IO.puts("Part1: #{result1}")

result2 =
  data
  |> Enum.filter(&AoC.overlap/1)
  |> Enum.count()

IO.puts("Part2: #{result2}")
